resources/media_controls.js

```diff
@@ -0,0 +1,354 @@
+      // For each transition, we check that the transition is possible and call
+      // the `onStateChange` handler.
+      for (let name in TRANSITIONS) {
+        if (!TRANSITIONS.hasOwnProperty(name)) {
+          continue;
+        }
+        this[name] = () => {
+          const from = this.state;
+
+          // Checks if the transition is valid in the current state.
+          if (!TRANSITIONS[name][from]) {
+            const error = `Transition "${name}" invalid for the current state "${from}"`;
+            console.error(error);
+            throw new Error(error);
+          }
+
+          const to = TRANSITIONS[name][from];
+
+          if (from == to) {
+            return;
+          }
+
+          // Transition to the next state.
+          this.state = to;
+          this.onStateChange(from);
+        };
+      }
+
+      // Set initial state.
+      this.state = this.media.paused ? PAUSED : PLAYING;
+      this.onStateChange(null);
+    }
+
+    // State change handler
+    onStateChange(from) {
+      this.render(from);
+    }
+
+    render(from = this.state) {
+      const isAudioOnly = this.media.localName == "audio";
+      if (!isAudioOnly) {
+        // XXX This should ideally use clientHeight/clientWidth,
+        //     but for some reason I couldn't figure out yet,
+        //     using it breaks layout.
+        this.root.style.height = this.media.videoHeight;
+        this.root.style.width = this.media.videoWidth;
+      }
+
+      // Error
+      if (this.state == ERRORED) {
+        //XXX render errored state
+        return;
+      }
+
+      if (this.state != from) {
+        // Play/Pause button.
+        const playPauseButton = this.elements.playPauseButton;
+        playPauseButton.classList.remove(from);
+        playPauseButton.classList.add(this.state);
+      }
+
+      // Progress.
+      const positionPercent = this.media.currentTime / this.media.duration * 100;
+      if (!isNaN(positionPercent) && positionPercent != Infinity) {
+        this.elements.progress.value = positionPercent;
+      } else {
+        this.elements.progress.value = 0;
+      }
+
+      // Current time and duration.
+      let currentTime = "0:00";
+      let duration = "0:00";
+      if (!isNaN(this.media.currentTime) && !isNaN(this.media.duration)) {
+        currentTime = formatTime(Math.round(this.media.currentTime * 1000));
+        duration = formatTime(Math.round(this.media.duration * 1000));
+      }
+      this.elements.positionDurationBox.show(currentTime, duration);
+
+      // Volume.
+      const volumeSwitchClass = this.media.muted || this.media.volume == 0 ? "muted" : "volumeup";
+      if (!this.elements.volumeSwitch.classList.contains(volumeSwitchClass)) {
+        this.elements.volumeSwitch.classList = "";
+        this.elements.volumeSwitch.classList.add(volumeSwitchClass);
+      }
+      const volumeLevelValue = Math.round((this.media.muted ? 0 : this.media.volume) * 100);
```

Comment:
To save a minimal amount of computational work, you could check if `this.media.muted` is `true` a little earlier.

```javascript
const volumeLevelValue = this.media.muted ? 0 : Math.round(this.media.volume * 100);
```